File: foreign/csharp/Iggy_SDK/Vsr/VsrNamespace.cs
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

using System.Buffers.Binary;
using Apache.Iggy.Enums;
using Apache.Iggy.Utils;

namespace Apache.Iggy.Vsr;

/// <summary>
///     Packs a stream/topic/partition triple into the <c>RequestHeader.namespace</c> word the server hashes
///     to a shard. Mirrors <c>core/binary_protocol/src/namespace.rs</c>; any drift routes writes to the
///     wrong shard.
/// </summary>
internal static class VsrNamespace
{
    internal const int MAX_STREAMS = 4096;
    internal const int MAX_TOPICS = 4096;
    internal const int MAX_PARTITIONS = 1_000_000;

    internal const int PARTITION_BITS = 20;
    internal const int TOPIC_BITS = 12;
    internal const int STREAM_BITS = 12;

    internal const int PARTITION_SHIFT = 0;
    internal const int TOPIC_SHIFT = PARTITION_SHIFT + PARTITION_BITS;
    internal const int STREAM_SHIFT = TOPIC_SHIFT + TOPIC_BITS;

    internal const ulong PARTITION_MASK = (1UL << PARTITION_BITS) - 1;
    internal const ulong TOPIC_MASK = (1UL << TOPIC_BITS) - 1;
    internal const ulong STREAM_MASK = (1UL << STREAM_BITS) - 1;

    /// <summary>
    ///     Reserved value routing a request to the cluster's single metadata consensus group. Sits above the
    ///     packed layout, so it can never collide with a packed namespace.
    /// </summary>
    internal const ulong METADATA_CONSENSUS_NAMESPACE = 1UL << 63;

    private const byte NumericIdKind = 1;
    private const byte StringIdKind = 2;
    private const byte ConsumerKind = 1;
    private const byte ConsumerGroupKind = 2;

    /// <summary>
    ///     The namespace a request header must carry, peeking into the classic payload for the ops the server
    ///     routes by partition. Named (string) stream or topic identifiers resolve to 0; the server resolves
    ///     them itself.
    /// </summary>
    internal static ulong ForRequest(int code, ReadOnlySpan<byte> payload, VsrOperation operation)
    {
        if (operation is VsrOperation.Register or VsrOperation.Logout)
        {
            return METADATA_CONSENSUS_NAMESPACE;
        }

        if (operation == VsrOperation.NonReplicated || operation.IsMetadata())
        {
            return 0;
        }

        return code switch
        {
            CommandCodes.SEND_MESSAGES_CODE => FromSendMessages(payload),
            CommandCodes.STORE_CONSUMER_OFFSET_CODE or CommandCodes.DELETE_CONSUMER_OFFSET_CODE
                or CommandCodes.STORE_CONSUMER_OFFSET_2_CODE or CommandCodes.DELETE_CONSUMER_OFFSET_2_CODE =>
                FromConsumerOffset(payload),
            CommandCodes.DELETE_SEGMENTS_CODE => FromDeleteSegments(payload),
            _ => throw VsrError.Exception(VsrError.FEATURE_UNAVAILABLE,
                $"Command code {code} cannot be routed to a partition namespace.")
        };
    }

    internal static ulong Pack(uint streamId, uint topicId, uint partitionId)
    {
        return ((streamId & STREAM_MASK) << STREAM_SHIFT)
               | ((topicId & TOPIC_MASK) << TOPIC_SHIFT)
               | ((partitionId & PARTITION_MASK) << PARTITION_SHIFT);
    }

    /// <summary>
    ///     <c>[metadata_len u32][stream][topic][partitioning][messages_count u32]</c>. Only the metadata
    ///     prefix is peeked; the message batch that follows is never parsed.
    /// </summary>
    private static ulong FromSendMessages(ReadOnlySpan<byte> payload)
    {
        var metadataLength = (int)ReadUInt32(payload, 0);
        if (metadataLength < 0 || payload.Length - 4 < metadataLength)
        {
            throw Malformed();
        }

        ReadOnlySpan<byte> metadata = payload.Slice(4, metadataLength);
        var position = 0;
        var streamId = ReadIdentifier(metadata, ref position);
        var topicId = ReadIdentifier(metadata, ref position);

        var kind = ReadByte(metadata, position);
        var length = ReadByte(metadata, position + 1);
        position += 2;
        if (kind != (byte)Partitioning.PartitionId)
        {
            throw VsrError.Exception(VsrError.FEATURE_UNAVAILABLE,
                "Under VSR the partition must be resolved by the client; balanced and message-key partitioning cannot be sent on the wire.");
        }

        if (length != 4 || metadata.Length < position + 4)
        {
            throw Malformed();
        }

        return FromPartition(streamId, topicId, ReadUInt32(metadata, position));
    }

    /// <summary>
    ///     <c>[consumer kind u8][consumer id][stream][topic][partition flag u8][partition u32]</c>, shared by
    ///     the store and delete consumer-offset requests and their v2 variants.
    /// </summary>
    private static ulong FromConsumerOffset(ReadOnlySpan<byte> payload)
    {
        var consumerKind = ReadByte(payload, 0);
        if (consumerKind is not (ConsumerKind or ConsumerGroupKind))
        {
            throw Malformed();
        }

        var position = 1;
        _ = ReadIdentifier(payload, ref position);
        var streamId = ReadIdentifier(payload, ref position);
        var topicId = ReadIdentifier(payload, ref position);

        var flag = ReadByte(payload, position);
        position += 1;

        if (flag == 0)
        {
            throw VsrError.Exception(VsrError.INVALID_IDENTIFIER,
                "Under VSR a consumer-offset request must carry an explicit partition id.");
        }

        if (payload.Length < position + 4)
        {
            throw Malformed();
        }

        return FromPartition(streamId, topicId, ReadUInt32(payload, position));
    }

    /// <summary>
    ///     <c>[stream][topic][partition u32][segments_count u32]</c>.
    /// </summary>
    private static ulong FromDeleteSegments(ReadOnlySpan<byte> payload)
    {
        var position = 0;
        var streamId = ReadIdentifier(payload, ref position);
        var topicId = ReadIdentifier(payload, ref position);
        if (payload.Length < position + 4)
        {
            throw Malformed();
        }

        return FromPartition(streamId, topicId, ReadUInt32(payload, position));
    }

    private static ulong FromPartition(uint? streamId, uint? topicId, uint partitionId)
    {
        if (streamId is null || topicId is null)
        {
            return 0;
        }

        Validate(streamId.Value, MAX_STREAMS);
        Validate(topicId.Value, MAX_TOPICS);
        Validate(partitionId, MAX_PARTITIONS);

        return Pack(streamId.Value, topicId.Value, partitionId);
    }

    private static void Validate(uint value, int exclusiveMax)
    {
        if (value >= exclusiveMax)
        {
            throw VsrError.Exception(VsrError.INVALID_IDENTIFIER,
                $"Identifier {value} is outside the packable namespace range (max {exclusiveMax - 1}).");
        }
    }

    /// <summary>
    ///     Reads an <c>[kind u8][len u8][value]</c> identifier, advancing <paramref name="position" />.
    ///     Returns the numeric value, or <c>null</c> for a named identifier the server has to resolve.
    ///     A kind or width the wire format does not define is corruption, not a named identifier: falling
    ///     back to <c>null</c> would route the request to namespace 0 instead of failing it.
    /// </summary>
    private static uint? ReadIdentifier(ReadOnlySpan<byte> payload, ref int position)
    {
        var kind = ReadByte(payload, position);
        var length = ReadByte(payload, position + 1);
        if (payload.Length < position + 2 + length)
        {
            throw Malformed();
        }

        uint? value = (kind, length) switch
        {
            (NumericIdKind, 4) => ReadUInt32(payload, position + 2),
            (StringIdKind, > 0) => null,
            _ => throw Malformed()
        };
        position += 2 + length;

        return value;
    }

    private static byte ReadByte(ReadOnlySpan<byte> payload, int offset)
    {
        if (payload.Length <= offset)
        {
            throw Malformed();
        }

        return payload[offset];
    }

    private static uint ReadUInt32(ReadOnlySpan<byte> payload, int offset)
    {
        if (payload.Length < offset + 4)
        {
            throw Malformed();
        }

        return BinaryPrimitives.ReadUInt32LittleEndian(payload.Slice(offset, 4));
    }

    private static Exception Malformed()
    {
        return VsrError.Exception(VsrError.INVALID_COMMAND, "Request payload is too short to resolve a namespace.");
    }
}

File: foreign/java/java-sdk/src/main/java/org/apache/iggy/client/async/tcp/vsr/VsrNamespace.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

package org.apache.iggy.client.async.tcp.vsr;

import io.netty.buffer.ByteBuf;
import org.apache.iggy.exception.IggyInvalidArgumentException;
import org.apache.iggy.exception.IggyServerException;

/**
 * Partition-plane namespace packing and payload peeking, mirroring
 * {@code core/binary_protocol/src/namespace.rs} and the request inspection in
 * {@code core/sdk/src/vsr.rs}.
 *
 * <p>The packed namespace routes a request to its partition's consensus
 * group: stream id in bits 32..43, topic id in bits 20..31, partition id in
 * bits 0..19. Control-plane requests (Register/Logout) use the metadata
 * sentinel {@code 1 << 63}; non-replicated and metadata ops use zero.
 */
final class VsrNamespace {

    static final long METADATA_CONSENSUS_NAMESPACE = 1L << 63;

    static final int MAX_STREAMS = 4096;
    static final int MAX_TOPICS = 4096;
    static final int MAX_PARTITIONS = 1_000_000;

    static final int STREAM_SHIFT = 32;
    static final int TOPIC_SHIFT = 20;

    private static final int SEND_MESSAGES_CODE = 101;
    private static final int STORE_CONSUMER_OFFSET_CODE = 121;
    private static final int DELETE_CONSUMER_OFFSET_CODE = 122;
    private static final int STORE_CONSUMER_OFFSET_2_CODE = 123;
    private static final int DELETE_CONSUMER_OFFSET_2_CODE = 124;
    private static final int DELETE_SEGMENTS_CODE = 503;

    private static final int IDENTIFIER_KIND_NUMERIC = 1;
    private static final int PARTITIONING_KIND_PARTITION_ID = 2;

    private VsrNamespace() {}

    static long namespaceFor(int operation, int commandCode, ByteBuf payload) {
        if (operation == VsrOperation.REGISTER || operation == VsrOperation.LOGOUT) {
            return METADATA_CONSENSUS_NAMESPACE;
        }
        if (operation == VsrOperation.NON_REPLICATED || VsrOperation.isMetadata(operation)) {
            return 0;
        }
        return switch (commandCode) {
            case SEND_MESSAGES_CODE -> fromSendMessages(payload);
            case STORE_CONSUMER_OFFSET_CODE,
                    DELETE_CONSUMER_OFFSET_CODE,
                    STORE_CONSUMER_OFFSET_2_CODE,
                    DELETE_CONSUMER_OFFSET_2_CODE -> fromConsumerOffset(payload);
            case DELETE_SEGMENTS_CODE -> fromDeleteSegments(payload);
            default -> throw IggyServerException.fromTcpResponse(VsrHeaders.ERROR_FEATURE_UNAVAILABLE, new byte[0]);
        };
    }

    /**
     * Payload: {@code [metadata_len:u32][stream id][topic id][partitioning]}.
     * Only explicit partition-id partitioning can be routed. Balanced and
     * key-based partitioning are resolved to a partition id by
     * {@code MessagesTcpClient} before the frame is encoded, so this check
     * is defense-in-depth for callers bypassing that resolution.
     */
    private static long fromSendMessages(ByteBuf payload) {
        ByteBuf in = payload.slice();
        if (in.readableBytes() < 4) {
            throw new IggyInvalidArgumentException("Send messages payload is missing its metadata section");
        }
        long metadataLength = in.readUnsignedIntLE();
        if (in.readableBytes() < metadataLength) {
            throw new IggyInvalidArgumentException("Send messages metadata section is truncated");
        }
        ByteBuf metadata = in.readSlice((int) metadataLength);
        Long streamId = readNumericIdentifier(metadata);
        Long topicId = readNumericIdentifier(metadata);
        requireReadable(metadata, 2, "Send messages payload is missing partitioning metadata");
        int partitioningKind = metadata.readUnsignedByte();
        int partitioningLength = metadata.readUnsignedByte();
        if (partitioningKind != PARTITIONING_KIND_PARTITION_ID || partitioningLength != 4) {
            throw IggyServerException.fromTcpResponse(VsrHeaders.ERROR_FEATURE_UNAVAILABLE, new byte[0]);
        }
        requireReadable(metadata, Integer.BYTES, "Send messages payload is missing its partition id");
        long partitionId = metadata.readUnsignedIntLE();
        return pack(streamId, topicId, partitionId);
    }

    /**
     * Payload: {@code [consumer kind:u8][consumer id][stream id][topic id]
     * [has_partition:u8][partition_id:u32]...}.
     */
    private static long fromConsumerOffset(ByteBuf payload) {
        ByteBuf in = payload.slice();
        requireReadable(in, 1, "Consumer offset payload is missing its consumer kind");
        in.skipBytes(1);
        skipIdentifier(in);
        Long streamId = readNumericIdentifier(in);
        Long topicId = readNumericIdentifier(in);
        requireReadable(in, 1, "Consumer offset payload is missing its partition presence flag");
        int hasPartition = in.readUnsignedByte();
        requireReadable(in, Integer.BYTES, "Consumer offset payload is missing its partition id");
        long partitionId = in.readUnsignedIntLE();
        if (hasPartition == 0) {
            throw new IggyInvalidArgumentException("Consumer offset requests require an explicit partition id");
        }
        return pack(streamId, topicId, partitionId);
    }

    /** Payload: {@code [stream id][topic id][partition_id:u32][segments_count:u32]}. */
    private static long fromDeleteSegments(ByteBuf payload) {
        ByteBuf in = payload.slice();
        Long streamId = readNumericIdentifier(in);
        Long topicId = readNumericIdentifier(in);
        requireReadable(in, Integer.BYTES * 2, "Delete segments payload is missing its partition id or segments count");
        long partitionId = in.readUnsignedIntLE();
        return pack(streamId, topicId, partitionId);
    }

    /** Named identifiers resolve server-side; they defer routing with namespace zero. */
    private static long pack(Long streamId, Long topicId, long partitionId) {
        if (streamId == null || topicId == null) {
            return 0;
        }
        validate(streamId, MAX_STREAMS, "stream id");
        validate(topicId, MAX_TOPICS, "topic id");
        validate(partitionId, MAX_PARTITIONS, "partition id");
        return (streamId << STREAM_SHIFT) | (topicId << TOPIC_SHIFT) | partitionId;
    }

    private static void validate(long value, int exclusiveMax, String field) {
        if (value >= exclusiveMax) {
            throw new IggyInvalidArgumentException(
                    "The " + field + " " + value + " exceeds the VSR namespace limit of " + (exclusiveMax - 1));
        }
    }

    private static Long readNumericIdentifier(ByteBuf in) {
        requireReadable(in, 2, "Identifier header is truncated");
        int kind = in.readUnsignedByte();
        int length = in.readUnsignedByte();
        requireReadable(in, length, "Identifier value is truncated");
        if (kind == IDENTIFIER_KIND_NUMERIC) {
            if (length != Integer.BYTES) {
                throw new IggyInvalidArgumentException("Numeric identifier must contain a 4-byte value");
            }
            return in.readUnsignedIntLE();
        }
        in.skipBytes(length);
        return null;
    }

    private static void skipIdentifier(ByteBuf in) {
        requireReadable(in, 2, "Identifier header is truncated");
        in.skipBytes(1);
        int length = in.readUnsignedByte();
        requireReadable(in, length, "Identifier value is truncated");
        in.skipBytes(length);
    }

    private static void requireReadable(ByteBuf in, int length, String message) {
        if (in.readableBytes() < length) {
            throw new IggyInvalidArgumentException(message);
        }
    }
}
